<launch>
    <!-- KOBUKI BASE ARGUMENTS -->
    <arg name="base"             default="$(env TURTLEBOT_BASE)"/>  <!-- create, roomba -->
    <arg name="battery"          default="$(env TURTLEBOT_BATTERY)"/>  <!-- /proc/acpi/battery/BAT0 in 2.6 or earlier kernels-->
    <arg name="stacks"           default="$(env TURTLEBOT_STACKS)"/>  <!-- circles, hexagons -->
    <arg name="3d_sensor"        default="$(env TURTLEBOT_3D_SENSOR)"/>  <!-- kinect, asus_xtion_pro -->
    <arg name="simulation"       default="$(env TURTLEBOT_SIMULATION)"/>
    <arg name="serialport"       default="$(env TURTLEBOT_SERIAL_PORT)"/> <!-- /dev/ttyUSB0, /dev/ttyS0 -->
    <arg name="robot_name"       default="$(env TURTLEBOT_NAME)"/>
    <arg name="robot_type"       default="$(env TURTLEBOT_TYPE)"/>

    <!-- CAMERA ARGUMENTS -->
    <arg name="camera"      default="camera"/>
    <arg name="publish_tf"  default="false"/>

    <!-- Factory-calibrated depth registration -->
    <arg name="depth_registration"              default="false"/>
    <arg     if="$(arg depth_registration)" name="depth" value="depth_registered" />
    <arg unless="$(arg depth_registration)" name="depth" value="depth" />

    <!-- Processing Modules -->
    <arg name="rgb_processing"                  default="true"/>
    <arg name="ir_processing"                   default="true"/>
    <arg name="depth_processing"                default="false"/>
    <arg name="depth_registered_processing"     default="true"/>
    <arg name="disparity_processing"            default="true"/>
    <arg name="disparity_registered_processing" default="true"/>
    <arg name="scan_processing"                 default="true"/>

    <!-- Worker threads for the nodelet manager -->
    <arg name="num_worker_threads" default="4" />

    <!-- Laserscan topic -->
    <arg name="scan_topic" default="scan"/>

    <!-- CMDVEL REPUBLISHER ARGUMENTS -->
    <arg name="x_lim" default="0.55"/>
    <arg name="z_lim" default="3.14159265359"/>

    <include file="$(find turtlebot_bringup)/launch/minimal.launch">
        <arg name="base" value="$(arg base)"/>
        <arg name="battery" value="$(arg battery)"/>
        <arg name="stacks" value="$(arg stacks)"/>
        <arg name="3d_sensor" value="$(arg 3d_sensor)"/>
        <arg name="simulation" value="$(arg simulation)"/>
        <arg name="serialport" value="$(arg serialport)"/>
        <arg name="robot_name" value="$(arg robot_name)"/>
        <arg name="robot_type" value="$(arg robot_type)"/>
    </include>
    <include file="$(find turtlebot_bringup)/launch/3dsensor.launch">
        <arg name="camera" value="$(arg camera)"/>
        <arg name="publish_tf" value="$(arg publish_tf)"/>
        <arg name="depth_registration" value="$(arg depth_registration)"/>
        <arg name="rgb_processing" value="$(arg rgb_processing)"/>
        <arg name="ir_processing" value="$(arg ir_processing)"/>
        <arg name="depth_processing" value="$(arg depth_processing)"/>
        <arg name="depth_registered_processing" value="$(arg depth_registered_processing)"/>
        <arg name="disparity_processing" value="$(arg disparity_processing)"/>
        <arg name="disparity_registered_processing" value="$(arg disparity_registered_processing)"/>
        <arg name="scan_processing" value="$(arg scan_processing)"/>
        <arg name="num_worker_threads" value="$(arg num_worker_threads)" />
        <arg name="scan_topic" value="$(arg scan_topic)"/>
    </include>

    <node pkg="uol_turtlebot_common" type="republish_cmd_vel.py" name="cmd_vel_republisher">
        <param name="x_lim" type="double" value="$(arg x_lim)"/>
        <param name="z_lim" type="double" value="$(arg z_lim)"/>
    </node>

    <!-- <node pkg="strands_webserver" type="strands_webserver" name="strands_webserver" output="screen">
        <param name="host_ip" value="$(optenv HOST_IP 127.0.0.1)" />
    </node> -->

    <!-- Map server -->
    <arg name="map_file" default="$(env TURTLEBOT_MAP_FILE)"/>
    <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)" />

    <arg name="initial_pose_x" default="0.0"/> <!-- Use 17.0 for willow's map in simulation -->
    <arg name="initial_pose_y" default="0.0"/> <!-- Use 17.0 for willow's map in simulation -->
    <arg name="initial_pose_a" default="0.0"/>
    <include file="$(find turtlebot_navigation)/launch/includes/amcl.launch.xml">
      <arg name="initial_pose_x" value="$(arg initial_pose_x)"/>
      <arg name="initial_pose_y" value="$(arg initial_pose_y)"/>
      <arg name="initial_pose_a" value="$(arg initial_pose_a)"/>
    </include>

    <arg name="custom_param_file" default="$(find turtlebot_navigation)/param/dummy.yaml"/>
    <include file="$(find turtlebot_navigation)/launch/includes/move_base.launch.xml">
      <arg name="custom_param_file" value="$(arg custom_param_file)"/>
    </include>


</launch>
